contracts/Rewards.sol

```diff
@@ -0,0 +1,102 @@
+// SPDX-License-Identifier: MIT
+
+pragma solidity ^0.8.0;
+
+import '@openzeppelin/contracts/utils/Address.sol';
+import '@openzeppelin/contracts/utils/Context.sol';
+import '@openzeppelin/contracts/token/ERC20/IERC20.sol';
+import './interfaces/IGovernance.sol';
+import './interfaces/IVault.sol';
+
+// Rewards contract.
+contract Rewards is Context {
+    IERC20 public rewardToken;
+    IVault public vault;
+    IGovernance public governance;
+    address public executor;
+    uint256 public rewardPerVote;
+    uint256 public allocated;
+
+    mapping(address => uint256) public locked;
+
+    event Deposited(address depositor, uint256 amount);
+    event Withdrawn(address withdrawer, uint256 amount);
+    event Claim(address claimer, uint256 proposalId, uint256 reward);
+    event Allocated(uint256 rewards);
+
+    modifier onlyGovernance() {
+        require(
+            _msgSender() == address(governance),
+            'Rewards:: onlyGovernance'
+        );
+        _;
+    }
+
+    modifier onlyExecutor() {
+        require(_msgSender() == executor, 'Rewards:: onlyExecutor');
+        _;
+    }
+
+    constructor(
+        IGovernance governance_,
+        address executor_,
+        IVault vault_,
+        IERC20 rewardToken_,
+        uint256 rewardPerVote_
+    ) {
+        governance = governance_;
+        executor = executor_;
+        rewardToken = rewardToken_;
+        vault = vault_;
+        rewardPerVote = rewardPerVote_;
+    }
+
+    function setVault(IVault vault_) external onlyExecutor {
+        vault = vault_;
+    }
+
+    function setRewardPerVote(uint256 rewardPerVote_) external onlyExecutor {
+        rewardPerVote = rewardPerVote_;
+    }
+
```

Comment:
The below allocate() function looks weird, it is not payable, no token transfer either, the executor just sends in a number that is incremented.

Can you review the purpose of this function and where it fits in the reward mechanism?

Reply:
You don't need getter functions to get rewardPerVote, you just call it .rewardPerVote() as its public storage variable and you can access it the same way in etherscan too.

Reply:
Not sure why to add require statement, if we set same value accidentally that's not a problem and most contract never add such checks for setters, if it doesn't hurt logic or security its better to avoid additional require.

Reply:
>Consider adding require(rewardPerVote =! rewardPerVote_) in the setter functions

🤔 ...that's a nice idea. If the value is already set at that, then the transaction reverts, saving the user cost. 
It would cover transaction ordering (race conditions) and honest user mistakes (not checking the value state)

Reply:
> as its public storage variable 

Yeah, I'm not the biggest fan of using public scope for state members (variables). My preference is  `private` or `internal` scoping and explicit `external` getters. 
Reasoning for the narrower scoping is encapsulation of variables within the context of contract operations (think composition of contracts interacting) and inheritance. 
With `external` getters providing marginally lower gas (at least they used to, untested recently) than the default `public` getters (https://ethereum.stackexchange.com/questions/19380/external-vs-public-best-practices).